server/storage-access-layer/src/main/java/ai/starwhale/mlops/storage/StorageAccessService.java

```diff
@@ -16,15 +16,107 @@
+import ai.starwhale.mlops.storage.s3.StorageAccessServiceS3;
 import java.io.IOException;
 import java.io.InputStream;
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 /**
  * provides file upload/ download /list services
  */
 public interface StorageAccessService {
 
+    class Registry {
+
+        private static final Map<String, Class<? extends StorageAccessService>> storageAccessServiceClassMap =
+                registerAll();
+
+        private static Map<String, Class<? extends StorageAccessService>> registerAll() {
+            Map<String, Class<? extends StorageAccessService>> ret = new HashMap<>();
+            ret.put("memory", StorageAccessServiceMemory.class);
+
+            ret.put("file", StorageAccessServiceFile.class);
+            ret.put("fs", StorageAccessServiceFile.class);
+
+            ret.put("s3", StorageAccessServiceS3.class);
+
+            ret.put("minio", StorageAccessServiceMinio.class);
+
+            ret.put("aliyun", StorageAccessServiceAliyun.class);
+            ret.put("oss", StorageAccessServiceAliyun.class);
+
+            ret.put("tencent", StorageAccessServiceQcloud.class);
+            ret.put("qcloud", StorageAccessServiceQcloud.class);
+
+            ret.put("baidu", StorageAccessServiceBos.class);
+            ret.put("bos", StorageAccessServiceBos.class);
+
+            ret.put("ksyun", StorageAccessServiceKsyun.class);
+            ret.put("ks3", StorageAccessServiceKsyun.class);
+
+            return ret;
+        }
+
+        public static Class<? extends StorageAccessService> getClassByType(String type) {
+            return storageAccessServiceClassMap.get(type.toLowerCase());
+        }
+
+        public static List<String> getUriSchemasByClass(Class<? extends StorageAccessService> clazz) {
+            return storageAccessServiceClassMap.entrySet().stream()
+                    .filter(entry -> entry.getValue() == clazz)
+                    .map(Entry::getKey)
+                    .collect(Collectors.toList());
+        }
+    }
+
+
+    Map<String, StorageAccessService> storageAccessServiceMap = new ConcurrentHashMap<>();
+
+
+    static StorageAccessService getMemoryStorageAccessService() {
+        return storageAccessServiceMap.computeIfAbsent("memory", key -> new StorageAccessServiceMemory());
+    }
+
+    static StorageAccessService getFileStorageAccessService(FsConfig config) {
+        return storageAccessServiceMap.computeIfAbsent("file-" + config.hashCode(),
+                key -> new StorageAccessServiceFile(config));
+    }
+
+    static StorageAccessService getS3LikeStorageAccessService(String type, S3Config config) {
+        return storageAccessServiceMap.computeIfAbsent(type.toLowerCase() + "-" + config.hashCode(), key -> {
```

Comment:
hash code is too easy to [collision](https://classes.engineering.wustl.edu/2019/spring/cse247/studios/9/)
Assuming that only the regions in the config are inconsistent, then whether the final result of the hash code is consistent depends on whether the regions will collide.

```java
public void testStringHashCodeCollision() {
    var hashTable = new HashMap<Integer, String>();
    // generate all string hash code with two char string
    for (int i = 'A'; i < 'z'; i++) {
        for (int j = 'A'; j < 'z'; j++) {
            var str = new String(new byte[] { (byte) i, (byte) j });
            var hashCode = str.hashCode();
            if (hashTable.containsKey(hashCode)) {
                System.out.println("hash code collision: " + str + " and " + hashTable.get(hashCode));
            }
            hashTable.put(hashCode, str);
        }
    }
}
```

part of the outputs

```
hash code collision: BB and Aa
hash code collision: BC and Ab
hash code collision: BD and Ac
hash code collision: BE and Ad
hash code collision: BF and Ae
hash code collision: BG and Af
hash code collision: BH and Ag
hash code collision: BI and Ah
hash code collision: BJ and Ai
hash code collision: BK and Aj
hash code collision: BL and Ak
hash code collision: BM and Al
hash code collision: BN and Am
hash code collision: BO and An
hash code collision: BP and Ao
hash code collision: BQ and Ap
hash code collision: BR and Aq
hash code collision: BS and Ar
hash code collision: BT and As
```

Reply:
Done. Change the key to a pair.